consensus/src/lib.rs

```diff
@@ -1,27 +1,1037 @@
+                                Address::new(config, address).expect("Failed to create address");
+                            Some((
+                                TxInput::new(
+                                    OutPointSourceId::Transaction(tx_id.clone()),
+                                    index as u32,
+                                    witness,
+                                ),
+                                TxOutput::new(
+                                    (output.get_value() - Amount::from(1)).unwrap(),
+                                    Destination::Address(receiver),
+                                ),
+                            ))
+                        } else {
+                            None
+                        }
+                    })
+                    .collect::<Vec<(TxInput, TxOutput)>>()
+            })
+            .unzip();
+
+        Block::new(
+            vec![Transaction::new(0, inputs, outputs, 0).expect("Failed to create transaction")],
+            if orphan {
+                Some(Id::new(&H256::random()))
+            } else {
+                Some(Id::new(&prev_block.get_id().get()))
+            },
+            time::get() as u32,
+            ConsensusData::None,
+        )
+        .expect("Error creating block")
+    }
+
+    #[test]
+    #[allow(clippy::eq_op)]
+    fn test_process_genesis_block_wrong_block_source() {
+        common::concurrency::model(|| {
+            // Genesis can't be from Peer, test it
+            let config = create_mainnet();
+            let storage = Store::new_empty().unwrap();
+            let mut consensus = Consensus::new(config.clone(), storage);
+
+            // process the genesis block
+            let block_source = BlockSource::Peer(0);
+            let result = consensus.process_block(config.genesis_block().clone(), block_source);
+            assert_eq!(result, Err(BlockError::InvalidBlockSource));
+        });
+    }
+
     #[test]
     #[allow(clippy::eq_op)]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
+    fn test_process_genesis_block() {
+        common::concurrency::model(|| {
+            // This test process only Genesis block
+            let config = create_mainnet();
+            let storage = Store::new_empty().unwrap();
+            let mut consensus = Consensus::new(config.clone(), storage);
+
+            // process the genesis block
+            let block_source = BlockSource::Local;
+            let block_index = consensus
+                .process_block(config.genesis_block().clone(), block_source)
+                .ok()
+                .flatten()
+                .unwrap();
+            assert_eq!(
+                consensus
+                    .blockchain_storage
+                    .get_best_block_id()
+                    .expect("Best block didn't found"),
+                Some(config.genesis_block().get_id())
+            );
+            assert_eq!(block_index.get_prev_block_id(), &None);
```

Comment:
Please also test that the premine transaction is accessible and that it's mainchain and spendable.